File: src/main/java/com/github/kindrat/cassandra/client/service/BackgroundTaskExecutor.java
package com.github.kindrat.cassandra.client.service;

import com.github.kindrat.cassandra.client.service.tasks.BackgroundTask;
import com.github.kindrat.cassandra.client.service.tasks.TaskStatistics;
import lombok.SneakyThrows;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

public class BackgroundTaskExecutor {
    private final Map<UUID, BackgroundTask> runningTasks = new HashMap<>();

    public UUID submit(BackgroundTask task) {
        UUID taskId = UUID.randomUUID();
        runningTasks.put(taskId, task);
        task.run();
        return taskId;
    }

    public Optional<TaskStatistics> findTaskStatistics(UUID taskId) {
        return Optional.ofNullable(runningTasks.get(taskId)).map(BackgroundTask::getStatistics);
    }

    @SneakyThrows
    public void stop(UUID taskId) {
        BackgroundTask task = runningTasks.remove(taskId);
        if (task != null) {
            task.dispose();
            while(!task.isDisposed()) {
                Thread.sleep(100);
            }
        }
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/service/BackgroundTaskExecutor.kt
package com.github.kindrat.cassandra.client.service

import com.github.kindrat.cassandra.client.service.tasks.BackgroundTask
import com.github.kindrat.cassandra.client.service.tasks.TaskStatistics
import java.util.*

class BackgroundTaskExecutor {
    private val runningTasks: MutableMap<UUID, BackgroundTask> = HashMap()

    fun submit(task: BackgroundTask): UUID {
        val taskId = UUID.randomUUID()
        runningTasks[taskId] = task
        task.run()
        return taskId
    }

    fun findTaskStatistics(taskId: UUID): TaskStatistics? {
        return runningTasks[taskId]?.statistics
    }

    fun stop(taskId: UUID) {
        runningTasks.remove(taskId)?.let {
            it.dispose()
            while (!it.isDisposed) {
                Thread.sleep(100)
            }
        }
    }
}